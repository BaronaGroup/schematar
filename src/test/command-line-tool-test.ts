import childProcess from 'child_process'
import fs from 'fs'

const isOutputFile = /\.schematar\.ts$/
const outputDir = __dirname + '/../../test-output/tool-test'

describe('command-line-tool-test', function () {
  beforeEach(() => prepareOutputDirectory(outputDir))

  it('can be used to set up typescript interface files', async function () {
    await invokeTool(['test-schemas/simple-schema.ts', 'test-schemas/named-schema.ts'])
    expect(fs.readdirSync(outputDir).sort()).toEqual(['named-schema.schematar.ts', 'simple-schema.schematar.ts'])
    expect(await fs.promises.readFile(outputDir + '/named-schema.schematar.ts', 'utf8')).toMatchSnapshot()
    expect(await fs.promises.readFile(outputDir + '/simple-schema.schematar.ts', 'utf8')).toMatchSnapshot()
  })

  it('supports globs', async function () {
    await invokeTool(['test-schemas/exp-*-schema.ts'])
    expect(fs.readdirSync(outputDir).sort()).toMatchSnapshot()
  })

  it('supports multiple globs', async function () {
    await invokeTool(['test-schemas/exp-*-schema.ts', 'test-schemas/no-*-schema.ts'])
    expect(fs.readdirSync(outputDir).sort()).toMatchSnapshot()
  })

  async function prepareOutputDirectory(dir: string) {
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir)
    }
    for (const file of fs.readdirSync(dir)) {
      const ffn = `${dir}/${file}`
      if (isOutputFile.test(ffn)) {
        fs.unlinkSync(ffn)
      }
    }
  }

  async function invokeTool(schemaArgs: string[]) {
    childProcess.spawnSync(
      'npx',
      ['ts-node', __dirname + '/../app', 'create-typescript-interfaces', outputDir, ...schemaArgs],
      {
        stdio: 'inherit',
        env: process.env,
        cwd: __dirname,
      }
    )
  }
})
