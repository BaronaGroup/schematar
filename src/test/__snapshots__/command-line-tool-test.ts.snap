// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`command-line-tool-test can be used to set up typescript interface files 1`] = `
"// Generated file, do not edit!

// tslint:disable
/* eslint-disable */
// @ts-ignore -- ignore possibly unused type parameters
export interface BananaBase<IDType, DateType>
{
  field: string
}
import {ObjectId} from 'mongodb'
export type BananaMongoose = BananaBase<ObjectId, Date>
export type BananaJSON = BananaBase<string, string>
export type BananaFluid = BananaBase<string | ObjectId, string | Date>"
`;

exports[`command-line-tool-test can be used to set up typescript interface files 2`] = `
"// Generated file, do not edit!

// tslint:disable
/* eslint-disable */
// @ts-ignore -- ignore possibly unused type parameters
export interface SimpleSchemaBase<IDType, DateType>
{
  field: string
}
import {ObjectId} from 'mongodb'
export type SimpleSchemaMongoose = SimpleSchemaBase<ObjectId, Date>
export type SimpleSchemaJSON = SimpleSchemaBase<string, string>
export type SimpleSchemaFluid = SimpleSchemaBase<string | ObjectId, string | Date>"
`;

exports[`command-line-tool-test supports globs 1`] = `
Array [
  "exp-context-schema.schematar.ts",
  "exp-name-schema.schematar.ts",
  "exp-schema-schema.schematar.ts",
]
`;

exports[`command-line-tool-test supports multiple globs 1`] = `
Array [
  "exp-context-schema.schematar.ts",
  "exp-name-schema.schematar.ts",
  "exp-schema-schema.schematar.ts",
  "no-extra-exports-schema.schematar.ts",
]
`;
